File: mahjong/ukeire_max.py
from __future__ import annotations

import random
import logging
from dataclasses import dataclass
from typing import Sequence

from mahjong.analyzer import DiscardAnalysis, analyze_discards
from mahjong.shanten import normal_shanten
from mahjong.tiles import TILE_NAMES, counts_to_tiles, empty_counts, tile_index, validate_counts


NUMBER_TILE_COUNT = 27
MIN_BEST_UKEIRE_TYPES = 6
MAX_SIMPLE_GAP = 4
LOGGER = logging.getLogger(__name__)


@dataclass(frozen=True)
class DifficultQuestionConfig:
    no_honors: bool = True
    require_iishanten: bool = True
    min_best_second_diff: int = 1
    max_best_second_diff: int = 4
    near_best_range: int = 5
    min_near_best_candidates: int = 3
    exclude_obvious_isolated_best_discard: bool = True
    prefer_middle_tiles: bool = True
    min_middle_tile_ratio: float = 0.6
    prefer_complex_shapes: bool = True
    max_difficult_attempts: int = 12
    relaxed_min_near_best_candidates: int = 2
    relaxed_min_middle_tile_ratio: float = 0.5


@dataclass(frozen=True)
class DifficultQuestionEvaluation:
    accepted: bool
    reject_reasons: tuple[str, ...]
    ranking: tuple[DiscardAnalysis, ...]
    best_second_diff: int | None
    near_best_candidate_count: int
    middle_tile_ratio: float
    complex_shape_score: int


DIFFICULT_QUESTION_CONFIG = DifficultQuestionConfig()


@dataclass(frozen=True)
class UkeireMaxQuestion:
    counts: tuple[int, ...]
    results: tuple[DiscardAnalysis, ...]
    best_discards: tuple[str, ...]

    @property
    def best_ukeire_tiles(self) -> int:
        return max(result.ukeire_tiles for result in self._best_results)

    @property
    def best_ukeire_types(self) -> int:
        return max(result.ukeire_types for result in self._best_results)

    @property
    def _best_results(self) -> tuple[DiscardAnalysis, ...]:
        best = set(self.best_discards)
        return tuple(result for result in self.results if result.discard in best)


def generate_ukeire_max_question(
    *,
    rng: random.Random | None = None,
    max_attempts: int = 20000,
    config: DifficultQuestionConfig = DIFFICULT_QUESTION_CONFIG,
) -> UkeireMaxQuestion:
    """Generate a 14-tile no-honor discard problem with iishanten ukeire answers."""
    generator = rng or random
    fallback: UkeireMaxQuestion | None = None
    best_scored: tuple[tuple[float, float], UkeireMaxQuestion] | None = None
    relaxed_fallback: UkeireMaxQuestion | None = None

    for attempt in range(max_attempts):
        counts = _random_no_honor_counts(generator)
        question = build_ukeire_max_question(counts)
        if question is None:
            continue
        if fallback is None:
            fallback = question

        if attempt >= config.max_difficult_attempts:
            if best_scored is not None:
                LOGGER.debug("Falling back to best scored ukeire max question after strict search limit.")
                return best_scored[1]
            if relaxed_fallback is not None:
                LOGGER.debug("Falling back to relaxed difficult ukeire max question after strict search limit.")
                return relaxed_fallback
            LOGGER.debug("Falling back to basic ukeire max question after strict search limit.")
            return fallback

        evaluation = evaluate_difficult_ukeire_question(question, config)
        if relaxed_fallback is None and evaluate_difficult_ukeire_question(question, _relaxed_config(config)).accepted:
            relaxed_fallback = question
        if not evaluation.accepted:
            continue

        score_key = (
            float(evaluation.complex_shape_score if config.prefer_complex_shapes else 0),
            evaluation.middle_tile_ratio if config.prefer_middle_tiles else 0.0,
        )
        if best_scored is None or score_key > best_scored[0]:
            best_scored = (score_key, question)
        if evaluation.complex_shape_score >= 3:
            return question

    if best_scored is not None:
        return best_scored[1]
    if fallback is not None:
        return fallback
    raise RuntimeError("Could not generate a ukeire max question.")


def build_ukeire_max_question(counts: Sequence[int]) -> UkeireMaxQuestion | None:
    validate_counts(counts, expected_total=14)
    if any(counts[index] for index in range(NUMBER_TILE_COUNT, len(TILE_NAMES))):
        return None

    results = tuple(analyze_discards(tuple(counts)))
    if min(result.after_discard_shanten for result in results) != 1:
        return None

    iishanten_results = tuple(result for result in results if result.after_discard_shanten == 1)
    if not iishanten_results:
        return None

    best_ukeire_tiles = max(result.ukeire_tiles for result in iishanten_results)
    best_results = tuple(result for result in iishanten_results if result.ukeire_tiles == best_ukeire_tiles)
    if not best_results:
        return None
    if min(result.ukeire_types for result in best_results) < MIN_BEST_UKEIRE_TYPES:
        return None

    return UkeireMaxQuestion(
        counts=tuple(counts),
        results=results,
        best_discards=tuple(sorted((result.discard for result in best_results), key=tile_index)),
    )


def evaluate_ukeire_max_answer(question: UkeireMaxQuestion, discards: Sequence[str] | str | None) -> bool:
    if discards is None:
        selected: set[str] = set()
    elif isinstance(discards, str):
        selected = {discards}
    else:
        selected = set(discards)
    return selected == set(question.best_discards)


def is_partial_ukeire_max_answer(question: UkeireMaxQuestion, discards: Sequence[str] | str | None) -> bool:
    if discards is None:
        selected: set[str] = set()
    elif isinstance(discards, str):
        selected = {discards}
    else:
        selected = set(discards)
    correct = set(question.best_discards)
    return bool(selected) and selected < correct


def filter_difficult_ukeire_question(
    question: UkeireMaxQuestion,
    config: DifficultQuestionConfig = DIFFICULT_QUESTION_CONFIG,
) -> bool:
    return evaluate_difficult_ukeire_question(question, config).accepted


def difficult_ukeire_debug_info(
    question: UkeireMaxQuestion,
    config: DifficultQuestionConfig = DIFFICULT_QUESTION_CONFIG,
) -> dict[str, object]:
    evaluation = evaluate_difficult_ukeire_question(question, config)
    return {
        "accepted": evaluation.accepted,
        "reject_reasons": evaluation.reject_reasons,
        "best_second_diff": evaluation.best_second_diff,
        "near_best_candidate_count": evaluation.near_best_candidate_count,
        "middle_tile_ratio": evaluation.middle_tile_ratio,
        "complex_shape_score": evaluation.complex_shape_score,
        "ranking": [
            {
                "discard": result.discard,
                "ukeire_types": result.ukeire_types,
                "ukeire_tiles": result.ukeire_tiles,
                "after_discard_shanten": result.after_discard_shanten,
            }
            for result in evaluation.ranking
        ],
    }


def evaluate_difficult_ukeire_question(
    question: UkeireMaxQuestion,
    config: DifficultQuestionConfig = DIFFICULT_QUESTION_CONFIG,
) -> DifficultQuestionEvaluation:
    counts = tuple(question.counts)
    ranking = get_discard_ukeire_ranking(question.results)
    middle_ratio = middle_tile_ratio(counts)
    complex_score = score_complex_shape(counts)
    reasons: list[str] = []

    if config.no_honors and _has_honors(counts):
        reasons.append("honors")

    if config.require_iishanten and not ranking:
        reasons.append("not_iishanten_after_discard")

    best_second_diff: int | None = None
    near_best_count = 0
    if ranking:
        best_tiles = ranking[0].ukeire_tiles
        near_best_count = sum(1 for result in ranking if best_tiles - result.ukeire_tiles <= config.near_best_range)
        if len(ranking) < 2:
            reasons.append("no_second_candidate")
        else:
            best_second_diff = best_tiles - ranking[1].ukeire_tiles
            if best_second_diff == 0:
                reasons.append("tied_best")
            elif not (config.min_best_second_diff <= best_second_diff <= config.max_best_second_diff):
                reasons.append("best_second_diff")

        if near_best_count < config.min_near_best_candidates:
            reasons.append("not_enough_near_best_candidates")

        if config.exclude_obvious_isolated_best_discard and is_obvious_isolated_tile(ranking[0].discard, counts):
            reasons.append("obvious_isolated_best_discard")

    if config.prefer_middle_tiles and middle_ratio < config.min_middle_tile_ratio:
        reasons.append("middle_tile_ratio")

    return DifficultQuestionEvaluation(
        accepted=not reasons,
        reject_reasons=tuple(reasons),
        ranking=ranking,
        best_second_diff=best_second_diff,
        near_best_candidate_count=near_best_count,
        middle_tile_ratio=middle_ratio,
        complex_shape_score=complex_score,
    )


def get_discard_ukeire_ranking(results: Sequence[DiscardAnalysis]) -> tuple[DiscardAnalysis, ...]:
    iishanten_results = [result for result in results if result.after_discard_shanten == 1]
    return tuple(
        sorted(
            iishanten_results,
            key=lambda result: (-result.ukeire_tiles, -result.ukeire_types, tile_index(result.discard)),
        )
    )


def is_obvious_isolated_tile(tile: str, counts: Sequence[int]) -> bool:
    index = tile_index(tile)
    if index >= NUMBER_TILE_COUNT or counts[index] != 1:
        return False

    suit_start = (index // 9) * 9
    rank = index % 9
    nearby_by_rank = {
        0: (1, 2),
        1: (0, 2, 3),
        7: (5, 6, 8),
        8: (6, 7),
    }
    nearby = nearby_by_rank.get(rank, (rank - 2, rank - 1, rank + 1, rank + 2))
    return all(counts[suit_start + nearby_rank] == 0 for nearby_rank in nearby)


def middle_tile_ratio(counts: Sequence[int]) -> float:
    total = sum(counts)
    if total == 0:
        return 0.0
    middle_tiles = 0
    for index, count in enumerate(counts[:NUMBER_TILE_COUNT]):
        rank = index % 9 + 1
        if 3 <= rank <= 7:
            middle_tiles += count
    return middle_tiles / total


def score_complex_shape(counts: Sequence[int]) -> int:
    score = 0
    suit_counts = counts[:NUMBER_TILE_COUNT]

    pair_count = sum(1 for count in suit_counts if count >= 2)
    triplet_count = sum(1 for count in suit_counts if count >= 3)
    if pair_count >= 2:
        score += pair_count - 1
    score += triplet_count * 2

    for suit_start in range(0, NUMBER_TILE_COUNT, 9):
        suit = list(counts[suit_start : suit_start + 9])
        score += _score_suit_complex_shape(suit)

    return score


def _random_no_honor_counts(generator: random.Random) -> tuple[int, ...]:
    counts = _random_no_honor_iishanten_counts(generator)
    while True:
        index = generator.randrange(NUMBER_TILE_COUNT)
        if counts[index] >= 4:
            continue
        counts[index] += 1
        return tuple(counts)


def _random_no_honor_iishanten_counts(generator: random.Random) -> list[int]:
    while True:
        counts = empty_counts()
        while sum(counts) < 13:
            index = generator.randrange(NUMBER_TILE_COUNT)
            if counts[index] >= 4:
                continue
            counts[index] += 1
        if normal_shanten(tuple(counts)) != 1:
            continue
        if _ukeire_type_count(counts) >= MIN_BEST_UKEIRE_TYPES:
            return counts


def _ukeire_type_count(counts: list[int]) -> int:
    target_shanten = normal_shanten(tuple(counts)) - 1
    ukeire_types = 0
    for index in range(NUMBER_TILE_COUNT):
        if counts[index] >= 4:
            continue
        drawn = counts.copy()
        drawn[index] += 1
        if normal_shanten(tuple(drawn)) <= target_shanten:
            ukeire_types += 1
    return ukeire_types


def _has_close_competition(results: tuple[DiscardAnalysis, ...], best_discards: tuple[str, ...]) -> bool:
    iishanten_results = [result for result in results if result.after_discard_shanten == 1]
    if len(iishanten_results) < 2:
        return False

    best_tiles = max(result.ukeire_tiles for result in iishanten_results)
    competitors = [
        result
        for result in iishanten_results
        if result.discard not in best_discards and best_tiles - result.ukeire_tiles <= MAX_SIMPLE_GAP
    ]
    return bool(competitors)


def _has_honors(counts: Sequence[int]) -> bool:
    return any(counts[index] for index in range(NUMBER_TILE_COUNT, len(TILE_NAMES)))


def _relaxed_config(config: DifficultQuestionConfig) -> DifficultQuestionConfig:
    return DifficultQuestionConfig(
        no_honors=config.no_honors,
        require_iishanten=config.require_iishanten,
        min_best_second_diff=config.min_best_second_diff,
        max_best_second_diff=config.max_best_second_diff + 1,
        near_best_range=config.near_best_range,
        min_near_best_candidates=config.relaxed_min_near_best_candidates,
        exclude_obvious_isolated_best_discard=config.exclude_obvious_isolated_best_discard,
        prefer_middle_tiles=config.prefer_middle_tiles,
        min_middle_tile_ratio=config.relaxed_min_middle_tile_ratio,
        prefer_complex_shapes=config.prefer_complex_shapes,
        max_difficult_attempts=config.max_difficult_attempts,
        relaxed_min_near_best_candidates=config.relaxed_min_near_best_candidates,
        relaxed_min_middle_tile_ratio=config.relaxed_min_middle_tile_ratio,
    )


def _score_suit_complex_shape(suit: list[int]) -> int:
    score = 0

    for rank in range(9):
        if suit[rank] < 3:
            continue
        if rank <= 6 and suit[rank + 1] and suit[rank + 2]:
            score += 2
        if rank >= 2 and suit[rank - 1] and suit[rank - 2]:
            score += 2
        if 0 < rank < 8 and suit[rank - 1] and suit[rank + 1]:
            score += 1

    for rank in range(7):
        if suit[rank] and suit[rank + 1] >= 2 and suit[rank + 2]:
            score += 2
        if suit[rank] >= 2 and suit[rank + 1] >= 2 and suit[rank + 2] >= 2:
            score += 3

    for rank in range(6):
        if suit[rank] and suit[rank + 1] >= 2 and suit[rank + 2] >= 2 and suit[rank + 3]:
            score += 3

    block_length = 0
    for count in suit:
        if count:
            block_length += 1
            if block_length >= 4:
                score += 1
        else:
            block_length = 0

    return score


def question_hand_text(question: UkeireMaxQuestion) -> str:
    return " ".join(counts_to_tiles(question.counts))

File: packages/mahjong-core/src/ukeireMax.ts
import { analyzeDiscards, bestDiscardsForReview, type DiscardAnalysis } from "./analyzer";
import { type Counts34, type Tile, countsToTiles, parseHand, tileIndex, validateCounts } from "./tiles";

const NUMBER_TILE_COUNT = 27;
const MIN_BEST_UKEIRE_TYPES = 6;
const DEFAULT_MAX_ATTEMPTS = 1000;

const UKEIRE_MAX_SEEDS = [
  "1223678m3455p899s",
  "456789m126667p35s",
  "5m3333678p455678s",
  "22266679m444457s",
  "117m4555p677899s",
  "24556678m233678p",
  "2333567m4589p456s",
  "111334m7899p3666s",
  "3m4446789p222334s",
  "7m134556788p2234s",
  "467778m223p23678s",
  "5567789m115p1345s",
  "155678m3458889p3s",
  "55999m4557p34567s",
  "13357899m567p678s",
  "234m25p333667899s",
  "5789m224455668p7s",
  "13555m124567p123s",
  "77m223444579p789s",
  "455m34567p112223s"
];

const DIFFICULT_UKEIRE_MAX_SEEDS = [
  "456789m126667p35s",
  "5m3333678p455678s",
  "24556678m233678p",
  "2333567m4589p456s",
  "3m4446789p222334s",
  "467778m223p23678s",
  "5567789m115p1345s",
  "55999m4557p34567s",
  "13357899m567p678s",
  "5789m224455668p7s",
  "13555m124567p123s",
  "77m223444579p789s",
  "455m34567p112223s",
  "566m3455p1113334s",
  "66m67899p123455s",
  "23m788p234566678s",
  "2333m2334468p345s",
  "44m23456p1334789s",
  "255567888m44s777s",
  "4m2345667p578999s"
];

export interface UkeireMaxQuestion {
  counts: Counts34;
  results: DiscardAnalysis[];
  bestDiscards: Tile[];
  bestUkeireTypes: number;
  bestUkeireTiles: number;
}

export type UkeireMaxAnswerState = "correct" | "partial" | "wrong";

export interface DifficultUkeireQuestionConfig {
  noHonors: boolean;
  requireIishanten: boolean;
  allowTiedBest: boolean;
  minBestDiscardCount: number;
  maxBestDiscardCount: number;
  minBestSecondDiff: number;
  maxBestSecondDiff: number;
  nearBestRange: number;
  minNearBestCandidates: number;
  excludeObviousIsolatedBestDiscard: boolean;
  preferMiddleTiles: boolean;
  minMiddleTileRatio: number;
  preferComplexShapes: boolean;
  requireTripletOrNearShape: boolean;
  maxDifficultAttempts: number;
  relaxedMinNearBestCandidates: number;
  relaxedMinMiddleTileRatio: number;
}

export interface DifficultUkeireQuestionEvaluation {
  accepted: boolean;
  rejectReasons: string[];
  ranking: DiscardAnalysis[];
  bestSecondDiff: number | null;
  nearBestCandidateCount: number;
  middleTileRatio: number;
  complexShapeScore: number;
}

export const DIFFICULT_UKEIRE_QUESTION_CONFIG: DifficultUkeireQuestionConfig = {
  noHonors: true,
  requireIishanten: true,
  allowTiedBest: false,
  minBestDiscardCount: 1,
  maxBestDiscardCount: 1,
  minBestSecondDiff: 1,
  maxBestSecondDiff: 4,
  nearBestRange: 5,
  minNearBestCandidates: 3,
  excludeObviousIsolatedBestDiscard: true,
  preferMiddleTiles: true,
  minMiddleTileRatio: 0.6,
  preferComplexShapes: true,
  requireTripletOrNearShape: false,
  maxDifficultAttempts: 80,
  relaxedMinNearBestCandidates: 2,
  relaxedMinMiddleTileRatio: 0.5
};

export const HARD_UKEIRE_QUESTION_CONFIG: DifficultUkeireQuestionConfig = {
  ...DIFFICULT_UKEIRE_QUESTION_CONFIG,
  allowTiedBest: true,
  minBestDiscardCount: 1,
  maxBestDiscardCount: 3,
  minBestSecondDiff: 0,
  maxBestSecondDiff: 4,
  minNearBestCandidates: 3,
  requireTripletOrNearShape: true,
  maxDifficultAttempts: 260
};

export function buildUkeireMaxQuestion(counts: Counts34): UkeireMaxQuestion | null {
  validateCounts(counts, 14);
  if (hasHonors(counts)) return null;

  const results = analyzeDiscards(counts).filter((result) => result.afterDiscardShanten === 1);
  if (results.length === 0) return null;

  const bestResults = bestDiscardsForReview(results);
  if (bestResults.length === 0) return null;

  const bestUkeireTypes = Math.max(...bestResults.map((result) => result.ukeireTypes));
  const bestUkeireTiles = Math.max(...bestResults.map((result) => result.ukeireTiles));
  if (bestUkeireTypes < MIN_BEST_UKEIRE_TYPES) return null;

  return {
    counts,
    results,
    bestDiscards: bestResults.map((result) => result.discard),
    bestUkeireTypes,
    bestUkeireTiles
  };
}

export function generateUkeireMaxQuestion(
  rng: () => number = Math.random,
  maxAttempts = DEFAULT_MAX_ATTEMPTS,
  config: DifficultUkeireQuestionConfig = DIFFICULT_UKEIRE_QUESTION_CONFIG
): UkeireMaxQuestion {
  for (const seed of shuffledDifficultSeeds(rng)) {
    const question = buildUkeireMaxQuestion(parseHand(seed));
    if (question && evaluateDifficultUkeireQuestion(question, config).accepted) return question;
  }

  let fallback: UkeireMaxQuestion | null = null;
  let relaxedFallback: UkeireMaxQuestion | null = null;
  let bestScored: { key: [number, number]; question: UkeireMaxQuestion } | null = null;

  for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
    const question = buildUkeireMaxQuestion(randomNumberHand(rng));
    if (!question) continue;
    if (!fallback) fallback = question;

    const evaluation = evaluateDifficultUkeireQuestion(question, config);
    if (!relaxedFallback && evaluateDifficultUkeireQuestion(question, relaxedConfig(config)).accepted) {
      relaxedFallback = question;
    }

    if (evaluation.accepted) {
      const key: [number, number] = [
        config.preferComplexShapes ? evaluation.complexShapeScore : 0,
        config.preferMiddleTiles ? evaluation.middleTileRatio : 0
      ];
      if (!bestScored || key[0] > bestScored.key[0] || (key[0] === bestScored.key[0] && key[1] > bestScored.key[1])) {
        bestScored = { key, question };
      }
      if (evaluation.complexShapeScore >= 3) return question;
    }

    if (attempt >= config.maxDifficultAttempts) {
      if (bestScored) return bestScored.question;
      if (relaxedFallback) return relaxedFallback;
      if (fallback) return fallback;
    }
  }

  for (const seed of shuffledSeeds(rng)) {
    const seeded = buildUkeireMaxQuestion(parseHand(seed));
    if (seeded) return seeded;
  }

  const finalFallback = buildUkeireMaxQuestion(parseHand("1455m2345677p678s"));
  if (!finalFallback) {
    throw new Error("No Ukeire MAX question could be generated.");
  }
  return finalFallback;
}

export function generateHardUkeireMaxQuestion(
  rng: () => number = Math.random,
  maxAttempts = DEFAULT_MAX_ATTEMPTS,
  config: DifficultUkeireQuestionConfig = HARD_UKEIRE_QUESTION_CONFIG
): UkeireMaxQuestion {
  for (const seed of shuffledDifficultSeeds(rng)) {
    const question = buildUkeireMaxQuestion(parseHand(seed));
    if (question && evaluateDifficultUkeireQuestion(question, config).accepted) return question;
  }

  let fallback: UkeireMaxQuestion | null = null;
  let relaxedFallback: UkeireMaxQuestion | null = null;
  let bestScored: { key: [number, number, number]; question: UkeireMaxQuestion } | null = null;

  for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
    const question = buildUkeireMaxQuestion(randomNumberHand(rng));
    if (!question) continue;

    const evaluation = evaluateDifficultUkeireQuestion(question, config);
    if (hasTripletOrNearShape(question.counts) && !fallback) fallback = question;

    const relaxed = evaluateDifficultUkeireQuestion(question, relaxedHardConfig(config));
    if (relaxed.accepted && !relaxedFallback) relaxedFallback = question;

    const score: [number, number, number] = [
      question.bestDiscards.length > 1 ? 2 : 0,
      evaluation.complexShapeScore,
      evaluation.middleTileRatio
    ];
    if (evaluation.accepted) {
      if (!bestScored || compareHardScore(score, bestScored.key) > 0) {
        bestScored = { key: score, question };
      }
      if (question.bestDiscards.length > 1 || evaluation.complexShapeScore >= 5) return question;
    }

    if (attempt >= config.maxDifficultAttempts) {
      if (bestScored) return bestScored.question;
      if (relaxedFallback) return relaxedFallback;
      if (fallback) return fallback;
    }
  }

  return generateUkeireMaxQuestion(rng, maxAttempts, relaxedConfig(config));
}

export function evaluateUkeireMaxAnswer(question: UkeireMaxQuestion, selected: Tile[]): UkeireMaxAnswerState {
  const answer = normalizeSelection(selected);
  const best = normalizeSelection(question.bestDiscards);
  if (answer.length === best.length && answer.every((tile) => best.includes(tile))) {
    return "correct";
  }
  if (answer.length > 0 && answer.length < best.length && answer.every((tile) => best.includes(tile))) {
    return "partial";
  }
  return "wrong";
}

export function toggleUkeireMaxSelection(selected: Tile[], tile: Tile): Tile[] {
  if (selected.includes(tile)) {
    return selected.filter((selectedTile) => selectedTile !== tile);
  }
  return [...selected, tile];
}

function normalizeSelection(tiles: Tile[]): Tile[] {
  return [...new Set(tiles)].sort((a, b) => tileIndex(a) - tileIndex(b));
}

export function evaluateDifficultUkeireQuestion(
  question: UkeireMaxQuestion,
  config: DifficultUkeireQuestionConfig = DIFFICULT_UKEIRE_QUESTION_CONFIG
): DifficultUkeireQuestionEvaluation {
  const ranking = getDiscardUkeireRanking(question.results);
  const middleRatio = middleTileRatio(question.counts);
  const complexScore = scoreComplexShape(question.counts);
  const reasons: string[] = [];

  if (config.noHonors && hasHonors(question.counts)) {
    reasons.push("honors");
  }

  if (config.requireIishanten && ranking.length === 0) {
    reasons.push("not_iishanten_after_discard");
  }

  let bestSecondDiff: number | null = null;
  let nearBestCandidateCount = 0;
  const bestDiscardCount = question.bestDiscards.length;
  if (ranking.length > 0) {
    const bestTiles = ranking[0]!.ukeireTiles;
    nearBestCandidateCount = ranking.filter((result) => bestTiles - result.ukeireTiles <= config.nearBestRange).length;

    if (bestDiscardCount < config.minBestDiscardCount || bestDiscardCount > config.maxBestDiscardCount) {
      reasons.push("best_discard_count");
    }

    if (ranking.length < 2) {
      reasons.push("no_second_candidate");
    } else {
      bestSecondDiff = bestTiles - ranking[1]!.ukeireTiles;
      if (bestSecondDiff === 0 && !config.allowTiedBest) {
        reasons.push("tied_best");
      } else if (!(config.minBestSecondDiff <= bestSecondDiff && bestSecondDiff <= config.maxBestSecondDiff)) {
        reasons.push("best_second_diff");
      }
    }

    if (nearBestCandidateCount < config.minNearBestCandidates) {
      reasons.push("not_enough_near_best_candidates");
    }

    if (config.excludeObviousIsolatedBestDiscard && isObviousIsolatedTile(ranking[0]!.discard, question.counts)) {
      reasons.push("obvious_isolated_best_discard");
    }
  }

  if (config.preferMiddleTiles && middleRatio < config.minMiddleTileRatio) {
    reasons.push("middle_tile_ratio");
  }

  if (config.requireTripletOrNearShape && !hasTripletOrNearShape(question.counts)) {
    reasons.push("not_triplet_or_near_shape");
  }

  return {
    accepted: reasons.length === 0,
    rejectReasons: reasons,
    ranking,
    bestSecondDiff,
    nearBestCandidateCount,
    middleTileRatio: middleRatio,
    complexShapeScore: complexScore
  };
}

export function getDiscardUkeireRanking(results: DiscardAnalysis[]): DiscardAnalysis[] {
  return results
    .filter((result) => result.afterDiscardShanten === 1)
    .slice()
    .sort((a, b) => b.ukeireTiles - a.ukeireTiles || b.ukeireTypes - a.ukeireTypes || tileIndex(a.discard) - tileIndex(b.discard));
}

export function filterDifficultUkeireQuestion(
  question: UkeireMaxQuestion,
  config: DifficultUkeireQuestionConfig = DIFFICULT_UKEIRE_QUESTION_CONFIG
): boolean {
  return evaluateDifficultUkeireQuestion(question, config).accepted;
}

export function difficultUkeireDebugInfo(question: UkeireMaxQuestion, config = DIFFICULT_UKEIRE_QUESTION_CONFIG) {
  const evaluation = evaluateDifficultUkeireQuestion(question, config);
  return {
    accepted: evaluation.accepted,
    rejectReasons: evaluation.rejectReasons,
    bestSecondDiff: evaluation.bestSecondDiff,
    nearBestCandidateCount: evaluation.nearBestCandidateCount,
    middleTileRatio: evaluation.middleTileRatio,
    complexShapeScore: evaluation.complexShapeScore,
    ranking: evaluation.ranking.map((result) => ({
      discard: result.discard,
      ukeireTypes: result.ukeireTypes,
      ukeireTiles: result.ukeireTiles,
      afterDiscardShanten: result.afterDiscardShanten
    }))
  };
}

export function isObviousIsolatedTile(tile: Tile, counts: Counts34): boolean {
  const index = tileIndex(tile);
  if (index >= NUMBER_TILE_COUNT || counts[index] !== 1) return false;

  const suitStart = Math.floor(index / 9) * 9;
  const rank = index % 9;
  const nearbyByRank = new Map<number, number[]>([
    [0, [1, 2]],
    [1, [0, 2, 3]],
    [7, [5, 6, 8]],
    [8, [6, 7]]
  ]);
  const nearby = nearbyByRank.get(rank) ?? [rank - 2, rank - 1, rank + 1, rank + 2];
  return nearby.every((nearbyRank) => counts[suitStart + nearbyRank] === 0);
}

export function middleTileRatio(counts: Counts34): number {
  const total = counts.reduce((sum, count) => sum + count, 0);
  if (total === 0) return 0;
  let middleTiles = 0;
  counts.slice(0, NUMBER_TILE_COUNT).forEach((count, index) => {
    const rank = (index % 9) + 1;
    if (3 <= rank && rank <= 7) middleTiles += count;
  });
  return middleTiles / total;
}

export function scoreComplexShape(counts: Counts34): number {
  const suitCounts = counts.slice(0, NUMBER_TILE_COUNT);
  const pairCount = suitCounts.filter((count) => count >= 2).length;
  const tripletCount = suitCounts.filter((count) => count >= 3).length;
  let score = pairCount >= 2 ? pairCount - 1 : 0;
  score += tripletCount * 2;

  for (let suitStart = 0; suitStart < NUMBER_TILE_COUNT; suitStart += 9) {
    score += scoreSuitComplexShape(counts.slice(suitStart, suitStart + 9));
  }
  return score;
}

export function hasTripletOrNearShape(counts: Counts34): boolean {
  if (counts.slice(0, NUMBER_TILE_COUNT).some((count) => count >= 3)) return true;

  for (let suitStart = 0; suitStart < NUMBER_TILE_COUNT; suitStart += 9) {
    const occupied = counts
      .slice(suitStart, suitStart + 9)
      .map((count, index) => count > 0 ? index : -1)
      .filter((index) => index >= 0);
    if (occupied.length < 3) continue;

    let componentLength = 1;
    let hasAdjacentOrOneGap = false;
    for (let i = 1; i < occupied.length; i += 1) {
      const gap = occupied[i]! - occupied[i - 1]!;
      if (gap <= 2) {
        componentLength += 1;
        if (gap <= 2) hasAdjacentOrOneGap = true;
        if (componentLength >= 5 && hasAdjacentOrOneGap) return true;
      } else {
        componentLength = 1;
        hasAdjacentOrOneGap = false;
      }
    }
  }

  return false;
}

function hasHonors(counts: Counts34): boolean {
  return counts.slice(NUMBER_TILE_COUNT).some((count) => count > 0);
}

function randomNumberHand(rng: () => number): Counts34 {
  const counts = Array(34).fill(0);
  while (countsToTiles(counts).length < 14) {
    const index = Math.floor(rng() * NUMBER_TILE_COUNT);
    if (counts[index] >= 4) continue;
    counts[index] += 1;
  }
  return counts;
}

function shuffledDifficultSeeds(rng: () => number): string[] {
  return shuffleStrings(DIFFICULT_UKEIRE_MAX_SEEDS, rng);
}

function relaxedConfig(config: DifficultUkeireQuestionConfig): DifficultUkeireQuestionConfig {
  return {
    ...config,
    maxBestSecondDiff: config.maxBestSecondDiff + 1,
    minNearBestCandidates: config.relaxedMinNearBestCandidates,
    minMiddleTileRatio: config.relaxedMinMiddleTileRatio
  };
}

function relaxedHardConfig(config: DifficultUkeireQuestionConfig): DifficultUkeireQuestionConfig {
  return {
    ...relaxedConfig(config),
    minBestSecondDiff: 0,
    requireTripletOrNearShape: config.requireTripletOrNearShape,
    maxBestDiscardCount: Math.max(config.maxBestDiscardCount, 3)
  };
}

function compareHardScore(left: [number, number, number], right: [number, number, number]): number {
  for (let index = 0; index < left.length; index += 1) {
    if (left[index]! !== right[index]!) return left[index]! - right[index]!;
  }
  return 0;
}

function scoreSuitComplexShape(suit: number[]): number {
  let score = 0;

  for (let rank = 0; rank < 9; rank += 1) {
    if (suit[rank]! < 3) continue;
    if (rank <= 6 && suit[rank + 1]! > 0 && suit[rank + 2]! > 0) score += 2;
    if (rank >= 2 && suit[rank - 1]! > 0 && suit[rank - 2]! > 0) score += 2;
    if (rank > 0 && rank < 8 && suit[rank - 1]! > 0 && suit[rank + 1]! > 0) score += 1;
  }

  for (let rank = 0; rank <= 6; rank += 1) {
    if (suit[rank]! > 0 && suit[rank + 1]! >= 2 && suit[rank + 2]! > 0) score += 2;
    if (suit[rank]! >= 2 && suit[rank + 1]! >= 2 && suit[rank + 2]! >= 2) score += 3;
  }

  for (let rank = 0; rank <= 5; rank += 1) {
    if (suit[rank]! > 0 && suit[rank + 1]! >= 2 && suit[rank + 2]! >= 2 && suit[rank + 3]! > 0) score += 3;
  }

  let blockLength = 0;
  for (const count of suit) {
    if (count > 0) {
      blockLength += 1;
      if (blockLength >= 4) score += 1;
    } else {
      blockLength = 0;
    }
  }

  return score;
}

function shuffledSeeds(rng: () => number): string[] {
  return shuffleStrings(UKEIRE_MAX_SEEDS, rng);
}

function shuffleStrings(values: string[], rng: () => number): string[] {
  const seeds = values.slice();
  for (let i = seeds.length - 1; i > 0; i -= 1) {
    const j = Math.floor(rng() * (i + 1));
    [seeds[i], seeds[j]] = [seeds[j]!, seeds[i]!];
  }
  return seeds;
}
